Mathlib/NumberTheory/NumberField/ClassNumber.lean

```diff
@@ -40,51 +41,103 @@ variable {K}
+    (h : ∀ ⦃p : ℕ⦄, p.Prime → p ≤ (4 / π) ^ nrComplexPlaces K *
+      ((finrank ℚ K)! / (finrank ℚ K) ^ (finrank ℚ K) * √|discr K|) →
+      ∀ (I : Ideal (𝓞 K)), I ∈ Ideal.primesOver (span {(p : ℤ)}) (𝓞 K) →
+      Submodule.IsPrincipal I) :
+    IsPrincipalIdealRing (𝓞 K) := by
+  refine RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_norm_le_of_isPrime <|
+    fun P hP hPNorm ↦ ?_
+  obtain ⟨p, hp⟩ := IsPrincipalIdealRing.principal <| under ℤ P.1
+  have hp0 : p ≠ 0 := fun h ↦ nonZeroDivisors.coe_ne_zero P <|
+    eq_bot_of_comap_eq_bot (R := ℤ) <| by simpa only [hp, submodule_span_eq, span_singleton_eq_bot]
+  have hpprime := (span_singleton_prime hp0).mp
+  rw [← submodule_span_eq, ← hp] at hpprime
+  have hle : algebraMap ℤ (𝓞 K) p ∈ P.1 := (mem_of_liesOver P.1 (under ℤ P.1) p).mp <|
+    hp ▸ Submodule.mem_span_singleton_self p
+  refine h (Int.prime_iff_natAbs_prime.mp (hpprime (hP.under _))) ?_ _ ⟨hP, ?_⟩
+  · refine le_trans (cast_le.mpr <| Nat.le_of_dvd ?_ (Int.ofNat_dvd_right.mp ?_)) hPNorm
+    · exact Nat.pos_of_ne_zero <| fun h ↦ nonZeroDivisors.coe_ne_zero P <| absNorm_eq_zero_iff.mp h
+    suffices (Algebra.norm ℤ (algebraMap ℤ (𝓞 K) p)) = p ^ (Module.finrank ℚ K) by
+      obtain ⟨i, -, hi⟩ := (dvd_prime_pow (hpprime (IsPrime.comap _)) _).mp
+        (this ▸ absNorm_dvd_norm_of_mem hle)
+      refine dvd_trans (dvd_pow_self p (fun h ↦ hP.ne_top (absNorm_eq_one_iff.mp ?_))) hi.dvd'
+      simp only [h, pow_zero, associated_one_iff_isUnit] at hi
+      exact ZMod.eq_one_of_isUnit_natCast hi
+    exact Rat.intCast_injective (by simp [Algebra.coe_norm_int, ← Algebra.norm_algebraMap])
+  · convert over_under P.1
+    rcases abs_choice p with h|h <;>
+    simp [hp, h]
+
 
 theorem _root_.RingOfIntegers.isPrincipalIdealRing_of_abs_discr_lt
     (h : |discr K| < (2 * (π / 4) ^ nrComplexPlaces K *
-      ((finrank ℚ K) ^ (finrank ℚ K) / (finrank ℚ K).factorial)) ^ 2) :
+      ((finrank ℚ K) ^ (finrank ℚ K) / (finrank ℚ K)!)) ^ 2) :
     IsPrincipalIdealRing (𝓞 K) := by
-  have : 0 < finrank ℚ K := finrank_pos -- Lean needs to know that for positivity to succeed
+  have : 0 < finrank ℚ K := finrank_pos -- Lean needs to know this for `positivity` to succeed
   rw [← Real.sqrt_lt (by positivity) (by positivity), mul_assoc, ← inv_mul_lt_iff₀' (by positivity),
     mul_inv, ← inv_pow, inv_div, inv_div, mul_assoc, Int.cast_abs] at h
-  rw [← classNumber_eq_one_iff, classNumber, Fintype.card_eq_one_iff]
-  refine ⟨1, fun C ↦ ?_⟩
-  obtain ⟨I, rfl, hI⟩ := exists_ideal_in_class_of_norm_le C
-  have : Ideal.absNorm I.1 = 1 := by
-    refine le_antisymm (Nat.lt_succ.mp ?_) (Nat.one_le_iff_ne_zero.mpr
-      (Ideal.absNorm_ne_zero_of_nonZeroDivisors I))
-    exact Nat.cast_lt.mp <| lt_of_le_of_lt hI h
-  rw [ClassGroup.mk0_eq_one_iff, Ideal.absNorm_eq_one_iff.mp this]
-  exact top_isPrincipal
+  refine RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_norm_le (fun I hI ↦ ?_)
+  convert top_isPrincipal
```

Comment:
that we pass from an `exat top_isPrincipal` to a `convert top_isPrincipal` seems a regression.

Reply:
I've changed it, but note that in this case `convert` is not (at least in my opinion) code smell. The use case is as when you have `h : a = 0` and your goal is `b = 0`: `convert h` leaves you with `a = b` but you never write `a` nor `b` (that can be very complicated, so a `suffice` may be annoying to write).